src/v/cluster/scheduling/allocation_state.cc

```diff
@@ -132,6 +132,37 @@ void allocation_state::update_allocation_nodes(
     }
 }
 
+void allocation_state::upsert_allocation_node(const model::broker& broker) {
+    auto it = _nodes.find(broker.id());
+
+    if (it == _nodes.end()) {
+        _nodes.emplace(
+          broker.id(),
+          std::make_unique<allocation_node>(
+            broker.id(),
+            broker.properties().cores,
+            broker.rack(),
+            _partitions_per_shard,
+            _partitions_reserve_shard0));
+    } else {
+        it->second->update_core_count(broker.properties().cores);
+        it->second->update_rack(broker.rack());
+        // node was added back to the cluster
+        if (it->second->is_removed()) {
```

Comment:
hmm so this is the node with the same id removed and then added, how can this happen?

Reply:
in allocation_state we do not remove old nodes, if a node was added back to the cluster with the same id it is marked as live again. This is releveant only when we reply controller log in topic table. i think this may be finally refactored to relaxe this dependency when node is removed.